main/src/ca/uwaterloo/flix/language/ast/Purity.scala

```diff
@@ -16,21 +16,131 @@
 
 package ca.uwaterloo.flix.language.ast
 
+import ca.uwaterloo.flix.util.InternalCompilerException
+
 sealed trait Purity
 
 /**
- * Represents the purity (or impurity) of an expression.
- */
+  * Represents the purity of an expression.
+  *
+  * - Pure expressions are treated as mathematical functions.
+  * - Impure expressions can use mutation or interact with the system.
+  * - Control-Impure expressions can do all the above but also use unhandled
+  *   algebraic effects.
+  *
+  * In terms of the set of expressions that is allowed under each effect the following holds:
+  * `e_pure ⊆ e_impure ⊆ e_control-impure`.
+  */
 object Purity {
 
   /**
-   * Represents a pure expression (i.e. an expression that cannot have side-effects).
-   */
+    * Represents a pure expression (i.e. an expression that cannot have side-effects).
+    */
   case object Pure extends Purity
 
   /**
-   * Represents an impure expression (i.e. an expression that could potentially have side-effects).
-   */
+    * Represents an impure expression (i.e. an expression that could potentially have side-effects).
+    */
   case object Impure extends Purity
 
+  /**
+    * Represents a control-impure expression (i.e. an expression that could potentially use effects like `do Print.print()`).
+    */
+  case object ControlImpure extends Purity
+
+  /**
+    * Returns true if `p` is a purity that can use algebraic effects
+    */
+  def canUseAlgebraicEffects(p: Purity): Boolean = p == ControlImpure
+
+  /**
+    * Returns the max effect of `p1` and `p2` according to this ordering:
+    * Pure < Impure < ControlImpure
+    */
+  def combine(p1: Purity, p2: Purity): Purity = (p1, p2) match {
+    case (Pure, Pure) => Pure
+    case (ControlImpure, _) => ControlImpure
+    case (_, ControlImpure) => ControlImpure
+    case (Impure, _) => Impure
```

Comment:
I'm not sure which line youre refering to since it shows me four lines, but controlpure has to be above impure. I've moved Pure down to the bottom though

Reply:
Actually, is there a reason not to spell out all cases? There are only 9 right?

Reply:
5 more places to typo perhaps, but I can do that. I could also make a `Purity => Int` function and use that